### DISCUSSION TITLE
Lesson 12 Part 4 | DecentralizedStablecoin.sol-src

### DISCUSSION THREAD

**Discussion by Rolko-u** [Cyfrin/foundry-full-course-cu#1338] [Q&A]
I'm trying to follow the course by typing the code in my vs code, but I get this error when I create my contract: 
```No arguments passed to the base constructor. Specify the arguments or mark "DecentralizedStableCoin" as abstract``` Here's my code: 
```
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.18;

import {ERC20Burnable, ERC20} from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";

contract DecentralizedStableCoin is ERC20Burnable, Ownable {
    constructor() ERC20("DecentralizedStableCoin", "DSC") {}
}
```
The error only appears when I set the ```Ownable``` as the parent contract. Need some help please

**Comment by ghost** [comment]
OpenZeppelin updated their smart contract library so apparently the new Ownable contract requires an initial owner as a constructor argument.

Inspect Ownable to see if that's the case. If it is, you can do:
- `Ownable(msg.sender)`, or 
- `constructor(address owner) ERC20("DecentralizedStableCoin", "DSC") Ownable(owner) {}`

**Reply by manu-acho** [reply]
I faced the same issue, how does this impact the DeployDSC.s.sol in the setup of the run() function. 
I have the contructor for DecentralizedStableCoin setup as
```solidity
constructor(address initialOwner) ERC20("DecentralizedStableCoin", "DSC") Ownable(initialOwner){}
```
and in the Deploy script
```solidity
contract DeployDSC is Script {
    function run() external returns (DecentralizedStableCoin, DSCEngine) {
        vm.startBroadcast();
        DecentralizedStableCoin dsc = new DecentralizedStableCoin(address(this));
        vm.stopBroadcast();
    }
}
```
I'm a bit worried I'd face issues with testing if this isn't setup correctly so any clues on how you managed this with the OpenZeppelin update would be much appreciated.

**Reply by PrintRH** [reply]
That's a good question looking ahead.

You're right that you will run into testing problems because the `initialOwner` address does not match the `DecentralizedStableCoin(address(this))` instance. What you can do is generate the public address of the `deployerKey` and pass it to the `dsc` variable (the section covering the part to pass `deployerKey` to `vm.startBroadcast(deployerKey)` will be covered by Patrick C. as you progressively watch through the videos).

Here's a code snippet of what that looks like:
```solidity
        vm.startBroadcast(deployerKey);
        address deployerPublicAddress = vm.addr(deployerKey);
        DecentralizedStableCoin dsc = new DecentralizedStableCoin(deployerPublicAddress);
        // rest of code...
        vm.stopBroadcast();
```

In the example above, you are then correctly matching the address that initiated the transaction which is also the `initialOwner` of `DecentralizedStableCoin` contract. With this setup, you should be good with running tests.

**Reply by manu-acho** [reply]
Thanks a lot!!
